numpy/_core/include/numpy/ndarraytypes.h

```diff
@@ -586,18 +586,32 @@ typedef struct _PyArray_Descr {
         /* unique-character representing this type */
         char type;
         /*
-         * '>' (big), '<' (little), '|'
-         * (not-applicable), or '=' (native).
+         * '>' (big), '<' (little), '|' (not-applicable), or '=' (native).
+         * Must be supported by all dtypes, only supporting '|' may break e.g.
+         * pickling.
          */
         char byteorder;
-        /* flags describing data type */
-        char flags;
+        /* Empty space, used to keep `type_num` aligned between 1.x and 2.x */
+        char _scratch1;
         /* number representing this type */
         int type_num;
+        /* flags describing data type */
+        npy_uint64 flags;
         /* element size (itemsize) for this type */
-        int elsize;
+        npy_intp elsize;
         /* alignment needed for this type */
-        int alignment;
+        npy_intp alignment;
+        /*
+         * Cached hash value (-1 if not yet computed).
+         */
+        npy_hash_t hash;
+
+        // TODO: The following need to be either moved into a "legacydescr"
```

Comment:
Overall, subclassing should make this easy enough in either case, but yes we could leave a `void *`
I don't like `c_metadata` here:  This is an object with `tp_dealloc` in it's type.  No need for another layer for freeing.

Reply:
Hmm, maybe subclassing then indeed is better, since the subclass can then be responsible for freeing whatever is needed (or this has to be `PyObject` so that one can `Py_XDECREF` it). Anyway, seems fine to leave it for later!